OrbitProducer/include/OrbitProducer/CaptureEventProducer.h

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_PRODUCER_CAPTURE_EVENT_PRODUCER_H_
+#define ORBIT_PRODUCER_CAPTURE_EVENT_PRODUCER_H_
+
+#include <thread>
+
+#include "absl/synchronization/mutex.h"
+#include "grpcpp/grpcpp.h"
+#include "producer_side_services.grpc.pb.h"
+
+namespace orbit_producer {
+
+// This abstract class offers the subclasses methods
+// to connect and communicate with a ProducerSideService.
+class CaptureEventProducer {
+ public:
+  // Pure virtual destructor, but still with definition (in .cpp file), makes this class abstract.
+  virtual ~CaptureEventProducer() = 0;
+
+  [[nodiscard]] bool IsCapturing() { return is_capturing_; }
+
+  // This method allows to specify how frequently a reconnection with the service should
+  // be attempted when the connection fails or is interrupted. The default is 5 seconds.
+  void SetReconnectionDelayMs(uint64_t ms) { reconnection_delay_ms_ = ms; }
+
+ protected:
+  // This method establishes the connection with ProducerSideService. If a connection fails or
+  // is interrupted, the class will keep trying to (re)connect, until ShutdownAndWait is called.
+  // Subclasses that extend this method by overriding it must also call the overridden method.
+  virtual void BuildAndStart(const std::shared_ptr<grpc::Channel>& channel);
+  // This method causes to disconnect from ProducerSideService or to stop trying to reconnect to it.
+  // Subclasses that extend this method by overriding it must also call the overridden method.
+  virtual void ShutdownAndWait();
+
+  // Subclasses can override this method to be notified of a request to start a capture.
+  virtual void OnCaptureStart();
+  // Subclasses can override this method to be notified of a request to stop the capture.
+  virtual void OnCaptureStop();
```

Comment:
Make them pure virtual for clarity, and you won't need virtual but implemented d-tor trick anymore.

Reply:
Done.